### ISSUE TITLE
More sophisticated reporting / queries based on tag-induced sets

### ISSUE THREAD

**Issue by apirogov** [almarklein/timetagger#244]
I started using timetagger a few days ago, its great and well-documented and I got it up and running with docker easily, so thanks for this great project!

That said, while I was looking for some simple approach to time tracking that I can host myself and easily export as CSV without loss of information, currently I feel that the reporting is rather primitive.

I love the tag-based nature, because you can put any time slice into multiple, overlapping sets - this sounds like something you can query in interesting ways (and possibly visualize nicely, but this is a different story). The current "report" apparently just groups together entries that have the exact same set of tags, which, kind-of, defeats the purpose and power of having free-form time tagging.

Are there any plans or ongoing work to use/present this information better?

I want something where I can choose a time range, fix a set of tags and then see how the percentage/total time distributes over all other tags, so I want to answer the question: **between T1 and T2** (e.g. *last week*), **given tag set X** (e.g. `#work #programming`), how much time/percent were tagged additionally with something else? (e.g. `#projectA` or `#projectB`).

**Example:**

If the time slices (numbered) in the chosen range were tagged as follows:

```
1: #talk to X about Y
2: #programming
3: #programming #projectA
4: #docs #projectA
5: #programming #projectA #testing
6: #programming #projectB #testing
7: #programming some feature
8: #meeting #projectA
```

And I want statistics for `#programming`, I expect to see a list like:
```
(no other tags): 2+7
#projectA: 3+5
#projectB: 6
#testing: 5+6
```
 
where the numbers on the right denote the slices to add up (and are divided by `sum(2,3,5,6,7)` for percentage).

If I would then want to know how much time in that range I was `#testing` in `#projectA`, I could simply click on `#projectA`, which would add it to the "reference set", so I would see for `#programming #projectA`:
```
(no other tags): 3
#testing: 5
```

and if I would next want to focus on `#projectA` in general, I could remove the `#programming` tag from the selection and see for `#projectA`:
```
#docs: 4
#programming: 3+5
#testing: 5
#meeting: 8
```

Of course there would be overlaps and it could add up to more than the total sum / 100%, and one could think of ways how to group this better and avoid this in cases where this is feasible (e.g. one could hide `#testing` in the last example and use `#programming #testing` instead, because `#testing` is fully subsumed), but even without any "deduplication logic" this would be very useful.

**Comment by almarklein** [comment]
To make sure we're on the same page: right now you can select a time range and select one or more tags. The shown groups include *all* records that have both tags (the AND operator). You want to be able to also select groups of records that have *any* of the selected records (the OR operator). 

Plus show percentages instead of hours.

**Comment by apirogov** [comment]
Percentage instead of hours in basically any view as on option would be quite convenient, yes.

But concerning the logic - no, I am not sure we mean the same thing.

In the current interface, the clicked/selected tags still should behave like AND and do the "pre-filtering"/narrowing (so as before, for the stats only time intervals within selected range and with ALL clicked tags are to be considered).

But currently, if I click on tag `A` and say, see
* A B D
* A B E
* A C D
* A C E

often what I *actually* want to know how much time I was doing `B/C/D/E` "overall" (given A). So when I click on tag `A`, what I want to see is:
* B (sum of all in current time range with A and B)
* C (sum of all in current range with A and C)
* ...

This might sound strange, but makes total sense if your tag system is a mix of contexts and activities, so that e.g.

* A is a project that I want to see some stats for (so A is the tag I click on),
* B/C are activities (such as communication, testing, refactoring, etc.)
* D/E are subprojects of A

So basically all kinds of combinations are possible.

Within project A, I want to see at a glance how much time I spent doing activity B, doing activity C, regardless of subproject/other tags combined with them. I don't want to add the numbers for "A B D" and "A B E" in my head, and I also don't want to click on B for the sum of "A B", go back, click on C for the sum of "A C" (which is basically an annoying work-around to get the numbers I want to see, using the fact that the total sum for the selected tags at the top). But currently I cannot conveniently get this without doing all this mental or mechanical extra work.

So I *do* want a kind of OR semantics variation, but not for the tag selection, but for what is shown to me in the list.

Let $I$ be the set of all tagged time intervals in time range $R$, $tags(x), len(x)$ be the set of tags and the length of a time interval $x \in I$, $T$ be the set of selected (clicked) tags and $J := \lbrace x \in I \mid T \subseteq tags(x) \rbrace \subseteq I$ the subset of intervals within R where all tags in $T$ are applied. Let $S := (\bigcup_{x \in J} tags(x)) \setminus T$ (all "additional" tags that appear in combination with T in timerange $R$).

Then in my proposed alternative view I would like to see one line $\forall a \in S$ that shows the total sum $\sum_{x \in J \cap a \in tags(x)} len(x)$ for that tag $a$.

**Comment by almarklein** [comment]
Thanks for the clarification!

Another attempt to phrase this in my own words: right now if you select one or more tags, the list of "subtags" is such that tag combinations are shown, and the total adds up to 100%. You would like the option that the list has one tag per item.

**Comment by apirogov** [comment]
Yes, exactly. I think "total per single tag" is generally interesting information to have that would be nice having "built-in", but if you don't think this is worth being treated "special" or the way I use tags is uncommon, its also fine :-)

Just wanted to propose this and see whether you think that this is something that might be a useful built-in feature, or I should build some quick ad-hoc thingy based on the API.

**Comment by almarklein** [comment]
Now the title makes perfect sense :)

Yes, I think having an option to show the overview this way makes sense 👍  Perhaps using a toggle button.

**Comment by apirogov** [comment]
Great, looking forward to it :)
